src/cli.rs

```diff
@@ -1,3 +1,558 @@
+        match args[2].as_str() {
+            "ls" | "list" => self.display_device_list(),
+            "new" => self
+                .find_device_factory_create_new_device(args)
+                .map_err(|e| self.display_message(e.as_str()))
+                .unwrap(),
+            "rm" | "remove" => self.remove_device(args),
+            _ => {
+                self.display_invalid_command();
+            }
+        }
+    }
+
+    fn display_device_list(&mut self) {
+        self.display_message("Device list:");
+        let devices = self.operations.list();
+        match devices {
+            Ok(devices) => {
+                for device in devices {
+                    self.display_message(&format!("Device: {}", device.get_name()));
+                }
+            }
+            Err(e) => {
+                self.display_message(&e);
+            }
+        }
+    }
+
+    fn find_device_factory_create_new_device(&mut self, args: Vec<String>) -> Result<(), String> {
+        if args.len() < 4 {
+            self.display_invalid_command();
+            return Ok(());
+        }
+        let device_key = args[3].as_str();
+        match self
+            .operations
+            .get_available_device_factories()
+            .iter()
+            .find(|&key| key.key == device_key)
+        {
+            Some(key) => self.create_new_device(key),
+            None => {
+                self.display_message("No such device configuration exists");
+                return Ok(());
+            }
+        }
+    }
+
+    fn create_new_device(&mut self, key: &DeviceFactoryKey) -> Result<(), String> {
+        self.display_message("Creating new device of type:");
+        let mut device_factory = self
+            .operations
+            .get_device_factory(key.key.clone())
+            .ok_or("No such device configuration exists")?;
+        while device_factory.has_next() {
+            let question_type = device_factory.get_question_type();
+            let question_statement = device_factory.get_question_statement();
+            let answer = self.ask_question(&question_type, &question_statement);
+            if let Some(device_factory) = Rc::get_mut(&mut device_factory) {
+                device_factory.set_question_answer(answer);
+            }
+        }
+        let device = device_factory
+            .build()
+            .map_err(|_| "Failed to build device")?;
+        self.operations
+            .add_device(device)
+            .map_err(|_| "Failed to add device")?;
+        self.display_message("Device created successfully");
+        Ok(())
+    }
+
+    fn remove_device(&mut self, args: Vec<String>) {
+        if args.len() < 4 {
+            self.display_invalid_command();
+            return;
+        }
+        let device_name = args[3].as_str();
+        match self.operations.remove_by_name(device_name.to_string()) {
+            Ok(_) => {
+                self.display_message("Removed device successfully");
+            }
```

Comment:
This too. Here you can even map the ok to the succes message, map the error to the error message: 

```rs
let message = self
            .operations
            .remove_by_name(device_name.to_string())
            .map(|_| "Removed device successfully")
            .unwrap_or("Failed to remove device");

        self.display_message(message);
```